services/QuillLMS/app/models/response.rb

```diff
@@ -0,0 +1,145 @@
+  belongs_to :response_prompt
+  belongs_to :response_question_type
+
+  validates_exclusion_of :correct, in: [nil]
+
+  # This is a list of keys from the old ConceptResults records
+  # that this model knows how to normalize.  Any keys that are
+  # fed into create_from_json that aren't from this list,
+  # will be stashed in a related ResponseExtraMetadata
+  # record
+  KNOWN_METADATA_KEYS = [
+    :answer,
+    :attemptNumber,
+    :correct,
+    :directions,
+    :lastFeedback,
+    :prompt,
+    :questionNumber,
+    :questionScore
+  ]
+
+  def self.create_from_json(data_hash)
+    response = init_from_json(data_hash)
+    response.save!
+    response
+  end
+
+  def self.init_from_json(data_hash)
+    data_hash = data_hash.deep_symbolize_keys
+
+    metadata = data_hash[:metadata]
+
+    response = new(
+      activity_session_id: data_hash[:activity_session_id],
+      answer: metadata[:answer],
+      concept_id: data_hash[:concept_id],
+      attempt_number: metadata[:attemptNumber],
+      correct: metadata[:correct],
+      question_number: metadata[:questionNumber],
+      question_score: metadata[:questionScore]
+    )
+
+    response.assign_normalized_text(data_hash)
+    response.assign_extra_metadata(metadata)
+
+    response
+  end
+
+  def self.bulk_create_from_json(data_hash_array)
+    data_hash_array.map { |data_hash| create_from_json(data_hash) }
+  end
+
+  def self.find_or_create_from_concept_result(concept_result)
+    return concept_result.response if concept_result.response
+
+    create_from_json(concept_result.as_json)
+  end
+
+  def self.parse_extra_metadata(metadata)
+    metadata.deep_symbolize_keys.except(*KNOWN_METADATA_KEYS).reject{ |_,v| v.nil? || (v.is_a?(Enumerable) && v.empty?) }
+  end
+
+  def legacy_format
+    {
+      id: id,
+      activity_classification_id: activity_session.activity.activity_classification_id,
+      activity_session_id: activity_session_id,
+      concept_id: concept.id,
+      question_type: response_question_type&.text,
+      metadata: legacy_format_metadata
+    }
+  end
+
+  def assign_normalized_text(data_hash)
+    metadata = data_hash[:metadata]
+
+    self.response_directions = ResponseDirections.find_or_create_by(text: metadata[:directions])
+    self.response_instructions = ResponseInstructions.find_or_create_by(text: metadata[:instructions])
+    self.response_previous_feedback = ResponsePreviousFeedback.find_or_create_by(text: metadata[:lastFeedback])
+    self.response_prompt = ResponsePrompt.find_or_create_by(text: metadata[:prompt])
+    self.response_question_type = ResponseQuestionType.find_or_create_by(text: data_hash[:question_type])
+  end
+
+  def assign_extra_metadata(metadata)
+    extra_metadata = self.class.parse_extra_metadata(metadata)
+
+    return if extra_metadata.empty?
+
+    self.extra_metadata = extra_metadata
+  end
+
+  private def legacy_format_metadata
+    legacy_format_base_metadata
+      .merge!(extra_metadata || {})
```

Comment:
This should use the non-destructive merge to maintain immutability
```suggestion
      .merge(extra_metadata || {})
```